cpp/src/arrow/compute/exec/forest_internal.h

```diff
@@ -21,15 +21,16 @@
 #include <utility>
 #include <vector>
 
-#include "arrow/dataset/visibility.h"
+#include "arrow/result.h"
+#include "arrow/status.h"
 
 namespace arrow {
-namespace dataset {
+namespace compute {
 
 /// A Forest is a view of a sorted range which carries an ancestry relation in addition
 /// to an ordering relation: each element's descendants appear directly after it.
 /// This can be used to efficiently skip subtrees when iterating through the range.
-class ARROW_DS_EXPORT Forest {
+class Forest {
```

Comment:
```suggestion
class ARROW_EXPORT Forest {
```

Reply:
The MSVC2015 linker gets really unhappy with this:

```
unity_0_cxx.cxx.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl arrow::compute::Forest::Forest(int,class std::function<bool __cdecl(int,int)>)" (__imp_??0Forest@compute@arrow@@QEAA@HV?$function@$$A6A_NHH@Z@std@@@Z) referenced in function "class arrow::compute::Forest __cdecl arrow::compute::MakeForest(class std::vector<struct arrow::compute::FileInfo,class std::allocator<struct arrow::compute::FileInfo> > *)" (?MakeForest@compute@arrow@@YA?AVForest@12@PEAV?$vector@UFileInfo@compute@arrow@@V?$allocator@UFileInfo@compute@arrow@@@std@@@std@@@Z)
unity_0_cxx.cxx.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: int __cdecl arrow::compute::Forest::size(void)const " (__imp_?size@Forest@compute@arrow@@QEBAHXZ) referenced in function "private: virtual void __cdecl arrow::compute::Forest_Visit_Test::TestBody(void)" (?TestBody@Forest_Visit_Test@compute@arrow@@EEAAXXZ)
unity_0_cxx.cxx.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: struct arrow::compute::Forest::Ref __cdecl arrow::compute::Forest::operator[](int)const " (__imp_??AForest@compute@arrow@@QEBA?AURef@012@H@Z) referenced in function "public: __cdecl arrow::compute::TestPathTree::TestPathTree(struct arrow::compute::Forest::Ref,class std::vector<struct arrow::compute::FileInfo,class std::allocator<struct arrow::compute::FileInfo> > const &)" (??0TestPathTree@compute@arrow@@QEAA@URef@Forest@12@AEBV?$vector@UFileInfo@compute@arrow@@V?$allocator@UFileInfo@compute@arrow@@@std@@@std@@@Z)
unity_0_cxx.cxx.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl arrow::compute::Forest::~Forest(void)" (__imp_??1Forest@compute@arrow@@QEAA@XZ) referenced in function "void __cdecl arrow::compute::ExpectForestIs(class std::vector<struct arrow::compute::FileInfo,class std::allocator<struct arrow::compute::FileInfo> >,class std::vector<struct arrow::compute::TestPathTree,class std::allocator<struct arrow::compute::TestPathTree> >)" (?ExpectForestIs@compute@arrow@@YAXV?$vector@UFileInfo@compute@arrow@@V?$allocator@UFileInfo@compute@arrow@@@std@@@std@@V?$vector@UTestPathTree@compute@arrow@@V?$allocator@UTestPathTree@compute@arrow@@@std@@@4@@Z)
release\arrow-compute-expression-test.exe : fatal error LNK1120: 4 unresolved externals
```